native_ops/fbw_op.cu

```diff
@@ -0,0 +1,613 @@
+                if (val <= pruning) {
+                    out << t << ' ' << (s - start_states[seq]) << ' ' << val
+                        << '\n';
+                }
+            }
+        }
+    }
+}
+
+void write_output_to_file(float *d_out, unsigned *d_seq_lens, float pruning,
+                          unsigned n_frames, unsigned n_seqs,
+                          unsigned n_emissions, unsigned batch_idx) {
+    std::vector<float> buffer(n_frames * n_seqs * n_emissions);
+    std::vector<unsigned> seq_lens(n_seqs);
+
+    HANDLE_ERROR(cudaMemcpy(buffer.data(), d_out, buffer.size() * sizeof(float),
+                            cudaMemcpyDeviceToHost));
+    HANDLE_ERROR(cudaMemcpy(seq_lens.data(), d_seq_lens,
+                            seq_lens.size() * sizeof(unsigned),
+                            cudaMemcpyDeviceToHost));
+
+    for (unsigned seq = 0u; seq < n_seqs; seq++) {
+        std::stringstream filename;
+        filename << "target.dump." << batch_idx << '.' << seq;
+        std::ofstream out(filename.str().c_str(),
+                          std::ios::out | std::ios::trunc);
+        for (unsigned t = 0u; t < n_frames; t++) {
+            if (t > 0u && t >= seq_lens[seq]) {
+                break;
+            }
+            for (unsigned e = 0u; e < n_emissions; e++) {
+                const float val =
+                    buffer[t * n_seqs * n_emissions + seq * n_emissions + e];
+                if (val <= pruning) {
+                    out << t << ' ' << e << ' ' << val << '\n';
+                }
+            }
+        }
+    }
+}
+
+std::vector<torch::Tensor> fbw_cuda(std::vector<torch::Tensor> torch_inputs) {
+    // am_scores, edges, weights, start_end_states, index, state_buffer
+    assert(torch_inputs.size() == 6);
+    Ndarray *am_scores = &torch_inputs[0];
+    Ndarray *edges = &torch_inputs[1];
+    Ndarray *weights = &torch_inputs[2];
+    Ndarray *start_end_states = &torch_inputs[3];
+    Ndarray *seq_lens = &torch_inputs[4];
+    Ndarray *state_buffer = &torch_inputs[5];
+
+    auto options = torch::TensorOptions().device(torch::kCUDA);
+    torch::Tensor out = torch::zeros_like(*am_scores, options);
+    torch::Tensor sum_output =
+        torch::zeros({am_scores->size(0), am_scores->size(1)}, options);
+
+    assert_cmp(Ndarray_DIMS(am_scores)[0], ==, Ndarray_DIMS(&out)[0]);
+    assert_cmp(Ndarray_DIMS(am_scores)[1], ==, Ndarray_DIMS(&out)[1]);
+    assert_cmp(Ndarray_DIMS(am_scores)[2], ==, Ndarray_DIMS(&out)[2]);
+    assert_cmp(Ndarray_DIMS(am_scores)[1], ==,
+               Ndarray_DIMS(start_end_states)[1]);
+    assert_cmp(Ndarray_DIMS(start_end_states)[0], ==, 2);
+
+    assert_cmp(Ndarray_DIMS(&sum_output)[0], ==, Ndarray_DIMS(am_scores)[0]);
+    assert_cmp(Ndarray_DIMS(&sum_output)[1], ==, Ndarray_DIMS(am_scores)[1]);
+
+    assert_cmp(Ndarray_DIMS(state_buffer)[0], ==, 2)
+
+        bool dump_alignment = false;
+    bool dump_output = false;
```

Comment:
Let's expose these options as parameters. Maybe inside a stuct? I.e.
```
struct DebugOptions {
    bool dump_output;
    ...
}
```

Reply:
Yes that's a good idea. Then one can also make this struct accessible as a Python class via pybind11.
But I'm wondering where it should be declared, such that it can be included in both fbw_op.cu and fbw_torch.cpp?